std/uni/package.d

```diff
@@ -7706,6 +7706,12 @@ public:
         return this.tupleof == other.tupleof;
     }
 
+    // Define a default toHash to allow AA usage
+    size_t toHash() const @trusted
+    {
+        return hashOf(slen_, hashOf(small_));
```

Comment:
AFAICT, this only works for the small-case, not for the heap-buffer case.

Reply:
The heap buffer data doesn't matter. The opEquals is just comparing the pointers.

Not saying this is what we *want* it to be, but that is what it is now.

Reply:
Yeah I know that the opEquals doesn't work in that case either, and that both implicit opEquals and toHash before v2.106 were most likely working in that case only too (IIRC, unions are by-default compared and hashed as a binary blob of the union's size).

Ideally, we'd be comparing/hashing the underlying bytes slice. But as the `[s]len_` fields seem to be the number of `dchar`s, a proper comparison/hash might have to rely on the `dchar` range returned by `opSlice()`.

Reply:
I don't actually know what the tupleof comparision does for the union...

Reply:
FWIW, I think the proper opEquals and toHash should be based on the opSlice.

Reply:
"it" being the default opEquals or being tupleof == tupleof?

Reply:
The new `tupleof` comparison. :)

Reply:
wow. I just tested it out. And this is pretty dangerous. For example, if one of the union members (that isn't valid) has an opEquals, then using the tupleof calls this, the default does a memcmp.

In this case of course, it's not a problem. But for sure this is not a good implementation.

Reply:
Oh, that reminds me of https://issues.dlang.org/show_bug.cgi?id=22171 - the same problem for autogenerated `__xtoHash`, which uses all union members.